include/mbgl/style/expression/expression.hpp

```diff
@@ -0,0 +1,270 @@
+            [&] (bool b) { return std::make_unique<LiteralExpression>(ctx.key(), b); },
+            [&] (auto f) {
+                auto number = numericValue<float>(f);
+                assert(number);
+                return std::make_unique<LiteralExpression>(ctx.key(), *number);
+            }
+        );
+    }
+    
+private:
+    optional<OutputValue> value;
+};
+
+class LambdaExpression : public Expression {
+public:
+    using Args = std::vector<std::unique_ptr<Expression>>;
+
+    LambdaExpression(std::string key,
+                    std::string name_,
+                    Args args_,
+                    type::Lambda type) :
+        Expression(key, type),
+        args(std::move(args_)),
+        name(name_)
+    {}
+    
+    template <class V>
+    static variant<CompileError, Args> parseArgs(const V& value, const ParsingContext& ctx) {
+        assert(isArray(value));
+        auto length = arrayLength(value);
+        Args args;
+        for(size_t i = 1; i < length; i++) {
+            const auto& arg = arrayMember(value, i);
+            auto parsedArg = parseExpression(arg, ParsingContext(ctx, i, {}));
+            if (parsedArg.template is<std::unique_ptr<Expression>>()) {
+                args.push_back(std::move(parsedArg.template get<std::unique_ptr<Expression>>()));
+            } else {
+                return parsedArg.template get<CompileError>();
+            }
+        }
+        return std::move(args);
+    }
+    
+protected:
+    Args args;
+private:
+    std::string name;
+};
+
+template <typename T, typename Rfunc>
+optional<OutputValue> evaluateBinaryOperator(float z,
+                                            const GeometryTileFeature& f,
+                                            EvaluationError& e,
+                                            const LambdaExpression::Args& args,
+                                            optional<T> initial,
+                                            Rfunc reduce)
+{
+    optional<T> memo = initial;
+    for(const auto& arg : args) {
+        auto argValue = arg->evaluate(z, f, e);
+        if (!argValue) return {};
+        if (!memo) memo = {argValue->get<T>()};
+        else memo = reduce(*memo, argValue->get<T>());
+    }
+    return {*memo};
+}
+
+class PlusExpression : public LambdaExpression {
+public:
+    PlusExpression(std::string key, Args args) :
+        LambdaExpression(key, "+", std::move(args),
+                        {type::Primitive::Number, {type::NArgs({type::Primitive::Number})}})
+    {}
+    
+    optional<OutputValue> evaluate(float zoom, const GeometryTileFeature& feature, EvaluationError& error) const override {
+        return evaluateBinaryOperator<float>(zoom, feature, error, args,
+            {}, [](float memo, float next) { return memo + next; });
+    }
+    
+    template <class V>
+    static ParseResult parse(const V& value, const ParsingContext& ctx) {
```

Comment:
All these `parse`s follow the same form, with only the expression type varying. That suggests it should be defined once as a template on the expression type.